File: auth-service/src/lib/ServiceDiscovery.ts
import consul, { Consul } from "consul";
import { hostname } from "os";
import { v4 as uuidv4 } from "uuid";

import { config } from "../config/consul";

/**
 * Singleton class for service discovery functionality.
 */
export class ServiceDiscovery {

    /**
     * Get the actual instance.
     */
    public static getInstance(): ServiceDiscovery {
        if (!this.instance) {
            this.instance = new ServiceDiscovery();
        }
        return this.instance;
    }

    private static instance: ServiceDiscovery;

    /**
     * Name of this service
     */
    public readonly serviceName: string;

    /**
     * Unique instance ID
     */
    public readonly instanceId: string;

    /**
     * Is the service registered?
     */
    private registered: boolean = false;

    /**
     * Interaval object for heartbeat
     */
    private interval: NodeJS.Timeout | null = null;

    /**
     * Consul client
     */
    private consul: Consul;

    /**
     * Heartbeat period in seconds
     */
    private heartbeatSeconds = 10;

    private constructor() {
        this.serviceName = config.clientName;
        this.instanceId = uuidv4();
        this.consul = new consul({
            host: config.consulHost,
        });
    }

    /**
     * Is this service registered to consul?
     */
    public isRegistered(): boolean {
        return this.registered;
    }

    /**
     * Register this instance into consul.
     */
    public registerService(errorHandler: (err: Error) => any): Promise<void> {
        const options: consul.Agent.Service.RegisterOptions = {
            name: this.serviceName,
            id: this.instanceId,
            address: hostname(),
            port: 4000,
            check: {
                ttl: `${this.heartbeatSeconds}s`,
            },
        };
        return new Promise<void>((resolve, reject) => {
            this.consul.agent.service.register(options, (err) => {
                if (err) {
                    return reject(err);
                }
                this.registered = true;
                this.registerHeartbeat(errorHandler);
                return resolve();
            });
        });
    }

    /**
     * Deregister this intance from consul.
     */
    public deregister(): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            this.deregisterHeartbeat();
            this.consul.agent.service.deregister(this.instanceId, (err) => {
                if (err) {
                    return reject(err);
                }
                this.registered = false;
                return resolve();
            });
        });
    }

    /**
     * Register the periodic heartbeat to consul
     */
    private registerHeartbeat(errorHandler: (err: Error) => any) {
        this.interval = setInterval(() => {
            this.consul.agent.check.pass(`service:${this.instanceId}`, (err) => {
                if (err) {
                    errorHandler(err);
                }
            });
        }, this.heartbeatSeconds * 500);
    }

    /**
     * Deregister the periodic heartbeat to consul
     */
    private deregisterHeartbeat() {
        if (this.interval !== null) {
            clearInterval(this.interval);
        }
        this.interval = null;
    }
}

File: user-service/app/service_discovery.py
import consul
import uuid
import platform
from app.config import consul_config, grpc_config
from app.utils import Interval
from typing import Callable

class ServiceDiscovery():
    """Service discovery class for publishing this service into Consul."""

    def __init__(self, failed_heartbeat_handler: Callable[[Exception], None] = None):
        """Construct a new service discovery instance.
        
        Keyword Arguments:

            failed_heartbeat_handler {Callable[[Exception], None]} -- Handler for failed heartbeats (default: {None})
        """
        super().__init__()
        self.registered = False
        self.consul = consul.Consul(host=consul_config["host"], port=consul_config["port"])
        self.service_name = consul_config["client_name"]
        self.instance_id = str(uuid.uuid4())
        self.heartbeat_seconds = 10
        self.healthy = True
        self.interval = None # type: Interval
        self.failed_heartbeat_handler = failed_heartbeat_handler

    def register(self):
        """Register the service into Consul."""
        self.consul.agent.service.register(
            name=self.service_name,
            service_id=self.instance_id,
            address=platform.node(),
            port=grpc_config["port"],
            check=consul.Check().ttl("{}s".format(self.heartbeat_seconds))
        )
        self.registered = True
        self.interval = Interval(8, self.pass_heartbeat)
        self.pass_heartbeat()

    def pass_heartbeat(self):
        """Pass another heartbeat (let Consul know that we are healthy)."""
        try:
            self.consul.agent.check.ttl_pass("service:{}".format(self.instance_id))
            self.healthy = True
        except Exception as e:
            self.healthy = False
            if self.failed_heartbeat_handler is not None:
                self.failed_heartbeat_handler(e)
            else:
                raise e

    def deregister(self):
        """Deregister the service from Consul."""
        self.consul.agent.service.deregister(self.instance_id)
        self.registered = False
        if self.interval is not None:
            self.interval.cancel()
            self.interval = None
